arch/arc/include/asm/atomic.h

```diff
@@ -31,13 +31,14 @@ static inline void atomic_##op(int i, atomic_t *v)			\
 	"	" #asm_op " %[val], %[val], %[i]	\n"		\
 	"	scond   %[val], [%[ctr]]		\n"		\
 	"	bnz     1b				\n"		\
-	: [val]	"=&r"	(val) /* Early clobber to prevent reg reuse */	\
+	: [val]	"=&r"	(val), /* Early clobber to prevent reg reuse */	\
+		"+m"	(v->counter)					\
```

Comment:
If you use `m` for ARCv3 in the above asm, you'll get an assembly error as the ARCv3 compiler will implicitly add `[]` around the memory argument.

Reply:
There will be no asm error here since that "+m" argument is unused and specified just to notify compiler about write operation into atomic. Anyway, after offline discussion with @claziss , current implementation has been reworked to address the following concerns:
- avoid unused argument in the inline assembly section
- add 'memory' clobber (unless it is explicitly clarified why it is not needed)

Here are the two options that could be used for inline assembly:
```
void atomic64_add(long a, atomic64_t *v)
{
  long val;
  long int *cnt = &v->counter;
 
 __asm__ __volatile__( "1:                \n"
            "    llockl   %0, [%1]     \n"
            "    addl     %0, %0, %2    \n"
            "    scondl   %0, [%1]     \n"
            "    bnz      1b        \n"
            : "=&r"(val), "+r" (cnt)
            : "r"(a)
            : "cc", "memory");
}
```
and
```
void atomic64_add(long a, atomic64_t *v)
{
  long val;

  __asm__ __volatile__( "1:                \n"
            "    llockl   %0, %1     \n"
            "    addl     %0, %0, %2    \n"
            "    scondl   %0, %1     \n"
            "    bnz      1b        \n"
            : "=&r"(val), "+m" (v->counter)
            : "r"(a)
            : "cc", "memory");
}
```
The second option has been selected. It turned out that "m" constraint was too general for llock/scond, generating the following error for several kernel functions: 
```
Error: inappropriate arguments for opcode 'llock' 
```
So "m" has been replaced by a more appropriate "+ATO" for ARCv2 and "+ATOMC" for ARCv3.